File: features/promo/tokopoints/src/main/java/com/tokopedia/tokopoints/view/customview/ServerErrorView.java
package com.tokopedia.tokopoints.view.customview;

import android.content.Context;
import android.content.res.TypedArray;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.widget.NestedScrollView;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.LayoutInflater;

import com.tokopedia.abstraction.common.utils.view.MethodChecker;
import com.tokopedia.tokopoints.R;

public class ServerErrorView extends NestedScrollView {

    private AppCompatImageView imageError;
    private AppCompatTextView tvTitleError;
    private AppCompatTextView tvLabelError;
    private RoundButton btnError;
    private CharSequence errorTitle;
    private CharSequence errorSubTitle;

    public ServerErrorView(@NonNull Context context) {
        super(context);
        init(null);
    }

    public ServerErrorView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(null);
    }

    public ServerErrorView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(null);
    }

    private void init(AttributeSet attrs) {
        readAttributes(attrs);
        inflateLayout();
        setDefaultValues();
    }

    private void readAttributes(AttributeSet attrs) {
        if (attrs != null) {

            TypedArray array = getContext().getTheme()
                    .obtainStyledAttributes(attrs, R.styleable.ServerErrorView, 0, 0);
            errorTitle = array.getString(R.styleable.ServerErrorView_errorTitle);
            errorSubTitle = array.getString(R.styleable.ServerErrorView_errorSubtitle);
            array.recycle();
        }
    }

    private void setDefaultValues() {
        if (TextUtils.isEmpty(errorTitle)) {
            errorTitle = getResources().getText(R.string.tp_label_server_error);
        }

        if (TextUtils.isEmpty(errorSubTitle)) {
            errorSubTitle = getResources().getText(R.string.tp_label_try_again);
        }
    }

    private void inflateLayout() {
        LayoutInflater.from(getContext()).inflate(R.layout.layout_tp_server_error, this, true);
        imageError = findViewById(R.id.img_error);
        tvTitleError = findViewById(R.id.text_title_error);
        tvLabelError = findViewById(R.id.text_label_error);
        btnError = findViewById(R.id.text_failed_action);
    }

    public void showErrorUi(boolean hasInternet) {

        int noConnectionImageId = R.drawable.ic_tp_toped_sorry;
        int buttonFontSize = getResources().getInteger(R.integer.tp_error_btn_medium);
        int buttonColor = MethodChecker.getColor(getContext(), com.tokopedia.design.R.color.transparent);
        int buttonFontColor = MethodChecker.getColor(getContext(), com.tokopedia.design.R.color.tkpd_main_green);

        if (!hasInternet) {

            noConnectionImageId = R.drawable.ic_tp_no_connection;
            buttonFontSize = getResources().getInteger(R.integer.tp_error_btn_large);

            buttonColor = MethodChecker.getColor(getContext(), com.tokopedia.design.R.color.bg_button_green_border_outline);

            buttonFontColor = MethodChecker.getColor(getContext(), com.tokopedia.design.R.color.white);
            errorTitle = getResources().getText(R.string.tp_no_internet_title);
            errorSubTitle = getResources().getText(R.string.tp_no_internet_label);
        }

        imageError.setImageResource(noConnectionImageId);

        btnError.setTextColor(buttonFontColor);
        btnError.setButtonColor(buttonColor);
        btnError.setTextSize(TypedValue.COMPLEX_UNIT_SP, buttonFontSize);

        tvTitleError.setText(errorTitle);
        tvLabelError.setText(errorSubTitle);
    }
}

File: features/promo/tokopoints/src/main/java/com/tokopedia/tokopoints/view/customview/ServerErrorView.kt
package com.tokopedia.tokopoints.view.customview

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.util.TypedValue
import android.view.LayoutInflater
import android.widget.ImageButton
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.appcompat.widget.Toolbar
import androidx.core.widget.NestedScrollView
import com.tokopedia.abstraction.common.utils.view.MethodChecker
import com.tokopedia.kotlin.extensions.view.hide
import com.tokopedia.kotlin.extensions.view.show
import com.tokopedia.tokopoints.R
import com.tokopedia.unifycomponents.ImageUnify
import com.tokopedia.unifycomponents.UnifyButton

class ServerErrorView : NestedScrollView {
    private var imageError: AppCompatImageView? = null
    private var tvTitleError: AppCompatTextView? = null
    private var tvLabelError: AppCompatTextView? = null
    private var btnError: RoundButton? = null
    private var errorTitle: CharSequence? = null
    private var errorSubTitle: CharSequence? = null
    private var errorBackArrow: UnifyButton? = null
    private var buttonFontSize: Int? = null
    private var buttonColor: Int? = null
    private var buttonFontColor: Int? = null
    private var containerToolbar: Toolbar?= null

    constructor(context: Context) : super(context) {
        init(null)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(null)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(null)
    }

    private fun init(attrs: AttributeSet?) {
        readAttributes(attrs)
        inflateLayout()
        setDefaultValues()
    }

    private fun readAttributes(attrs: AttributeSet?) {
        if (attrs != null) {
            val array = context.theme
                    .obtainStyledAttributes(attrs, R.styleable.ServerErrorView, 0, 0)
            errorTitle = array.getString(R.styleable.ServerErrorView_errorTitle)
            errorSubTitle = array.getString(R.styleable.ServerErrorView_errorSubtitle)
            array.recycle()
        }
    }

    private fun setDefaultValues() {
        if (TextUtils.isEmpty(errorTitle)) {
            errorTitle = resources.getText(R.string.tp_label_server_error)
        }
        if (TextUtils.isEmpty(errorSubTitle)) {
            errorSubTitle = resources.getText(R.string.tp_label_try_again)
        }
        buttonFontSize = resources.getInteger(R.integer.tp_error_btn_large)
        buttonColor = MethodChecker.getColor(context, com.tokopedia.unifyprinciples.R.color.Unify_G500)
        buttonFontColor = MethodChecker.getColor(context, com.tokopedia.unifyprinciples.R.color.Unify_N0)
    }

    private fun inflateLayout() {
        LayoutInflater.from(context).inflate(R.layout.layout_tp_server_error, this, true)
        imageError = findViewById(R.id.img_error)
        tvTitleError = findViewById(R.id.text_title_error)
        tvLabelError = findViewById(R.id.text_label_error)
        btnError = findViewById(R.id.text_failed_action)
        errorBackArrow = findViewById(R.id.iv_back_error)
        containerToolbar = findViewById(R.id.top_container)

        errorBackArrow?.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_tp_arrow_back,0,0,0)
    }

    fun showErrorUi(hasInternet: Boolean, error: Int = 0) {
        if (error == 1) {
            containerToolbar?.show()
        }
        var noConnectionImageId = com.tokopedia.globalerror.R.drawable.unify_globalerrors_500
        if (!hasInternet) {
            noConnectionImageId = com.tokopedia.globalerror.R.drawable.unify_globalerrors_connection
            errorTitle = resources.getText(R.string.tp_no_internet_title)
            errorSubTitle = resources.getText(R.string.tp_no_internet_label)
        }
        imageError?.setImageResource(noConnectionImageId)
        buttonFontColor?.let { btnError?.setTextColor(it) }
        buttonColor?.let { btnError?.setButtonColor(it) }
        buttonFontSize?.toFloat()?.let { btnError?.setTextSize(TypedValue.COMPLEX_UNIT_SP, it) }
        tvTitleError?.text = errorTitle
        tvLabelError?.text = errorSubTitle
    }

    fun setErrorButtonClickListener(onClickListener: OnClickListener?) {
        btnError?.setOnClickListener(onClickListener)
    }

    fun setErrorBackButtonClickListener(onClickListener: OnClickListener?) {
        errorBackArrow?.setOnClickListener(onClickListener)
    }
}